working/333 - shared memory multithreading/proposal.md

```diff
@@ -477,317 +493,53 @@ shared global state.
-  ShareableList<E> toShareable() =>
-    switch (this) {
-      final ShareableList<Shareable?> shareable => shareable,
-      _ => ShareableList<E>.from(this),
-    };
-}
-
-extension ToShareableSet<E extends Shareable?> on Set<E> {
-  ShareableSet<E> toShareable() => /* ... */
-}
-
-extension ToShareableMap<K extends Shareable?,
-                         V extends Shareable?> on Map<K, V> {
-  ShareableMap<K, V> toShareable() => /* ... */
-}
-```
-
-### `SendPort` semantics
-
-`SendPort` is extended with a new method to which allows sending `Shareable`
-values by reference without copying:
-
-```dart
-abstract interface class SendPort {
-    /// Sends an asynchronous [message] through this send port, to its
-    /// corresponding [ReceivePort].
-    ///
-    /// The message is passed by reference to the receiver without
-    /// copying.
-    ///
-    /// If sender and receiver do not share the same code then
-    /// an [IllegalArgument] exception is thrown.
-    void share(Shareable message);
-}
-```
-
-### `ShareableBox<T>`
-
-In some situations we might need to put a non-shareable value inside a shareable
-type. This is okay as long as we can guarantee that this value will only be
-accessed within the isolate it originally belonged to.
-
-```dart
-abstract interface class ShareableBox<T> implements Shareable {
-    factory ShareableBox(T value);
-    T get value;
-}
-
-abstract interface class MutableShareableBox<T> implements ShareableBox<T> {
-    factory MutableShareableBox(T value);
-    set value(T newValue);
-}
-```
-
-> [!NOTE]
->
-> It is possible to implement `ShareableBox<T>` on top of `Expando` but for
-> efficiency reasons we might want to provide a built-in implementation
->
-> ```dart
-> class _MutableShareableBoxImpl<T> implements MutableShareableBox<T> {
->     final _lock = Lock();
->     var _token = _AccessToken();
->     static final _values = Expando<AccessToken, (T,)>();
->
->     T get value => _lock.runLocked(() =>
->        switch (_values[_token]) {
->          (final v,) => v,
->          _ => throw StateError("not owned by current isolate"),
->        };
->     });
+> [!CAUTION]
 >
->     set value(T v) {
->        _lock.runLocked(() {
->          // If we don't own this box then reset access token so that
->          // expandos in the other isolate can get cleared.
->          if (_values[_token] == null) {
->      	    _token = _AccessToken();
->          }
->          _values[_token] = v;
->        });
->     }
-> }
+> Shared field reads and writes are _atomic_ for reference types, but other
+> than that there are no implicit synchronization, locking or strong memory
+> barriers associated with shared fields. Possible executions in terms of
+> observed values will be specified by the Dart's [memory model](#memory-models)
+> which I propose to model after JavaScript's and Go's: **program which is
+> free of data races will execute in a sequentially consistent manner**.
```

Comment:
Maybe it would make sense to describe here whether we'd allow or disallow initializers (or maybe it's somewhere in this markdown?)

* Do we allow initializer expressions?
* When concurrent mutators access it in parallel would we execute it N times and a undefined one will win?

Reply:
Documented that we expected atomic initialization where only one thread will run the initializer.